File: Luogu/P4777/chinese_remainder_theorem.cpp
#include <bits/stdc++.h>
using namespace std;

#define SIZE 100010

long long int cstArr[SIZE], modArr[SIZE];
int equNum;

long long int fastMul(long long int a, long long int n, long long int mod) {
    long long int ret = 0;
    a %= mod;
    while (n > 0) {
        if (n & 1)
            ret = (ret + a) % mod;
        a = (a << 1) % mod;
        n >>= 1;
    }
    return ret;
}

long long int extEuclid(long long int a, long long int b, long long int & x, long long int & y) {
    if (b == 0) {
        x = 1, y = 0;
        return a;
    }
    long long int gcd = extEuclid(b, a % b, x, y), tmp = x;
    x = y;
    y = tmp - y * (a / b);
    return gcd;
}

long long int crt() {
    long long int modProd = modArr[0], ans = cstArr[0];
    for (int i = 1; i < equNum; i++) {
        long long int x, y, a = modProd, b = modArr[i];
        long long int c = (cstArr[i] - ans % b + b) % b;
        long long int gcd = extEuclid(a, b, x, y), bg = b / gcd;

        if (c % gcd != 0)
            return -1;

        long long int cntAns = fastMul(x, c / gcd, bg);
        ans += modProd * cntAns;
        modProd *= bg;
        ans = (ans % modProd + modProd) % modProd;
    }

    return (ans % modProd + modProd) % modProd;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(0); cout.tie(0);
    cin >> equNum;
    for (int i = 0; i < equNum; i++)
        cin >> modArr[i] >> cstArr[i];
    cout << crt() << '\n';
    return 0;
}

File: Nowcoder/890D/chinese_remainder_theorem.java
import java.io.*;
import java.util.*;
import java.math.*;
 
public class Main {
    static BigInteger mods[] = new BigInteger[1010];
    static BigInteger csts[] = new BigInteger[1010];
 
    static BigInteger x, y;
 
    public static BigInteger extEuclid(BigInteger a, BigInteger b) {
        if (b.compareTo(BigInteger.ZERO) == 0) {
            x = BigInteger.ONE; y = BigInteger.ZERO;
            return a;
        }
        BigInteger gcd = extEuclid(b, a.mod(b));
        BigInteger tmp = x;
        x = y; y = tmp.subtract(y.multiply(a.divide(b)));
        return gcd;
    }
 
    public static BigInteger crt(int equNum) {
        BigInteger modProd = mods[0], ans = csts[0];
        for (int i = 1; i < equNum; i++) {
            BigInteger a = modProd, b = mods[i];
            BigInteger c = (csts[i].subtract(ans.mod(b)).add(b)).mod(b);
            BigInteger gcd = extEuclid(a, b);
            BigInteger bg = b.divide(gcd);
 
            if (c.mod(gcd).compareTo(BigInteger.ZERO) != 0) {
                return BigInteger.valueOf(-1);
            }
 
            BigInteger cntAns = c.divide(gcd).mod(bg).multiply(x).mod(bg).add(bg).mod(bg);
            ans = ans.add(modProd.multiply(cntAns));
            modProd = modProd.multiply(bg);
        }
 
        return ans.mod(modProd).add(modProd).mod(modProd);
    }
 
    public static void main(String[] args) {
        Scanner scn = new Scanner(System.in);
        int equNum = scn.nextInt(); BigInteger lim = scn.nextBigInteger();
        for (int i = 0; i < equNum; i++) {
            mods[i] = scn.nextBigInteger();
            csts[i] = scn.nextBigInteger();
        }
 
        BigInteger ret = crt(equNum);
        if (ret.compareTo(BigInteger.ZERO) == -1) {
            System.out.println("he was definitely lying");
        } else if (ret.compareTo(lim) == 1) {
            System.out.println("he was probably lying");
        } else {
            System.out.println(ret);
        }
    }
}